app/src/forms/rbac/service.js

```diff
@@ -0,0 +1,440 @@
+
+      await FormRoleUser.query(trx).patchAndFetchById(obj.id, update);
+      await trx.commit();
+      const result = await service.read(obj.id);
+      return result;
+    } catch (err) {
+      if (trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  read: async (id) => {
+    return FormRoleUser.query()
+      .findById(id)
+      .allowGraph('[form, role, user]')
+      .withGraphFetched('[form, role, user]')
+      .throwIfNotFound();
+  },
+
+  delete: async (id) => {
+    return FormRoleUser.query()
+      .deleteById(id)
+      .throwIfNotFound();
+  },
+
+  createUser: async (data) => {
+    let trx;
+    try {
+      trx = await transaction.start(User.knex());
+
+      const obj = Object.assign({}, data);
+      obj.id = uuidv4();
+
+      await User.query(trx).insert(obj);
+      await trx.commit();
+      const result = await service.read(obj.id);
+      return result;
+    } catch (err) {
+      if (trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  readUser: async (id) => {
+    return User.query()
+      .findById(id)
+      .throwIfNotFound();
+  },
+
+  updateUser: async (id, data) => {
+    let trx;
+    try {
+      const obj = await service.readUser(id);
+      trx = await transaction.start(User.knex());
+
+      const update = {
+        keycloakId: data.keycloakId,
+        username: data.username,
+        fullName: data.fullName,
+        email: data.email,
+        firstName: data.firstName,
+        lastName: data.lastName};
+
+      await User.query(trx).patchAndFetchById(obj.id, update);
+      await trx.commit();
+      const result = await service.readUser(id);
+      return result;
+    } catch (err) {
+      if (trx) await trx.rollback();
+      throw err;
+    }
+  },
+
+  getCurrentUser: async (access_token) => {
+    // if this user does not exists, add...
+    // return user details including form access...
+    let user = await User.query()
+      .first()
+      .where('keycloakId', access_token.content.sub.toString());
+
+    if (!user) {
+      // add to the system.
+      user = await service.createUser({
+        keycloakId: access_token.content.sub.toString(),
+        fullName: access_token.content.name.toString(),
+        username: access_token.content.preferred_username.toString(),
+        firstName: access_token.content.given_name.toString(),
+        lastName: access_token.content.family_name.toString(),
+        email: access_token.content.email.toString()
+      });
+    } else {
+      // what if name or email changed?
+      user = await service.updateUser(user.id,{
+        keycloakId: access_token.content.sub.toString(),
+        fullName: access_token.content.name.toString(),
+        username: access_token.content.preferred_username.toString(),
+        firstName: access_token.content.given_name.toString(),
+        lastName: access_token.content.family_name.toString(),
+        email: access_token.content.email.toString()
+      });
```

Comment:
Since the passed in object for keycloakId, fullName... email is used in both branches of the conditional, that should be specified as a const outside of it to remove redundancy.
```js
const userData = { keycloakId: ... };
if (!user) {
    user = await service.createUser(userData);
} else {
    user = await service.updateUser(user.id, userData);
}
```

Reply:
done